File: src/_cpp/ml/Animation.cpp
//
//  IslandDefense
//
//  Created by Vladimir Tolmachev on 27.09.14.
//
//
#include "Animation.h"
#include <stdio.h>
#include "cocos2d.h"
#include <stdarg.h>
#include "ImageManager.h"
#include "ml/common.h"
using namespace cocos2d;

cocos2d::Animation* createAnimation(const std::string & path, int firstIndex, int lastIndex, const std::string & fileExt, float duration)
{
	auto buildIndex = [firstIndex, lastIndex]( int index )
	{
		std::string result = intToStr( index );
		auto numbers = std::max( intToStr( firstIndex ).size(), intToStr( lastIndex ).size() );
		while( result.size() < numbers )
		{
			result = "0" + result;
		}
		return result;
	};
	std::vector<std::string>indexes;
	if( lastIndex>=firstIndex )
	{
		for(int i=firstIndex; i<=lastIndex; ++i )
			indexes.push_back( buildIndex( i ) );
	}
	else
	{
		for(int i=lastIndex; i<=firstIndex; ++i )
			indexes.push_back( buildIndex( i ) );
		std::reverse(indexes.begin(), indexes.end());
	}
	return createAnimation(path, indexes, fileExt, duration);
}

Animation* createAnimation(const std::string & path, std::vector<std::string> indexes, const std::string & fileExt, float duration)
{
	std::vector<std::string> files;
	for( auto index : indexes )
	{
		files.push_back( path + index + fileExt );
	}
	return createAnimation( files, duration );
}

cocos2d::Animation* createAnimation(std::vector<std::string> textures, float duration)
{
	cocos2d::Vector<cocos2d::SpriteFrame*> frames;
	for( auto texturePath : textures )
	{
		auto frame = ImageManager::shared().spriteFrame( texturePath );
		if( frame )
		{
			frames.pushBack(frame );
			continue;
		}
		//
		cocos2d::Texture2D * texture = Director::getInstance()->getTextureCache()->addImage(texturePath);
		if(!texture)
		{
#if USE_CHEATS == 1
			std::string message = "cannot create animation. Path [" + texturePath + "].";
			log( "%s", message.c_str() );
			MessageBox( message.c_str(), "Animation not created" );
#endif
			continue;
		}
		cocos2d::Rect rect;
		rect.size = texture->getContentSize();

		frames.pushBack( cocos2d::SpriteFrame::create( texturePath.c_str(), rect) );
	}
	float delay = duration / frames.size();

	return cocos2d::Animation::createWithSpriteFrames(frames, delay);
}

//void computePointsByRadius(std::vector<cocos2d::Point> & out, float radius, unsigned countPoints, float startAngleInDegree)
//{
//	float delta = static_cast<float>(M_PI) * 2.0f / countPoints;
//	float startAngleInRadian = startAngleInDegree * static_cast<float>(M_PI) / 180.f;
//	out.resize(countPoints);
//	for( unsigned i=0; i<countPoints; ++i )
//	{
//		float angle = startAngleInRadian + delta * i;
//		out[i].x = radius * cos(angle);
//		out[i].y = radius * sin(angle);
//	}
//}

File: src/_eu/ml/Animation.js
/******************************************************************************
 * Copyright 2014-2016 Vladimir Tolmachev
 * Copyright 2016 Visionarity AG
 * Vladimir Tolmachev and Visionarity AG have unlimited commercial
 * licenses for commercial use and customization
 *
 * Author: Vladimir Tolmachev (tolm_vl@hotmail.com)
 * Ported C++ to Javascript: Visionarity AG / Vladimir Tolmachev
 * Project: Island Defense (JS)
 * If you received the code not from the author, please contact us
 ******************************************************************************/

//Define namespace
var EU = EU || {};

EU.Animation = {

    createAnimation: function( a, b, c, d, e) {
        "use strict";
        if ( e !== undefined ) {
            this.createAnimation_n(a,b,c,d,e)
        }
        else if ( d !== undefined ) {
            this.createAnimation_i(a,b,c,d)
        }
        else if ( b !== undefined ) {
            this.createAnimation_t(a,b)
        }
    },

    createAnimation_n: function (path, firstIndex, lastIndex, fileExt, duration) {
        "use strict";
        var buildIndex = function (firstIndex, lastIndex, index) {
            var result = index.toString();
            var numbers = Math.max(firstIndex.toString().length, lastIndex.toString().length);
            while (result.length < numbers) {
                result = "0" + result;
            }
            return result;
        };
        var indexes = [];
        if (lastIndex >= firstIndex) {
            for (var i = firstIndex; i <= lastIndex; ++i)
                indexes.push(buildIndex(firstIndex, lastIndex, i));
        }
        else {
            EU.assert(0);
            //for( i=lastIndex; i<=firstIndex; ++i )
            //    indexes.push( buildIndex( firstIndex, lastIndex, i ) );
            //std::reverse(indexes.begin(), indexes.end());
        }
        return this.createAnimation(path, indexes, fileExt, duration);
    },

    createAnimation_i: function (path, indexes, fileExt, duration) {
        "use strict";
        var files = [];
        for (var i = 0; i < indexes; ++i )
        {
            files.push(path + indexes[i] + fileExt);
        }
        return this.createAnimation(files, duration);
    },
    createAnimation_t: function (textures, duration) {
        "use strict";
        var frames = [];
        for (var i = 0; i < textures.length; ++i) {
            var texturePath = textures[i];
            var frame = EU.ImageManager.getSpriteFrame(texturePath);
            if (frame) {
                frames.push(frame);
                continue;
            }
            var texture = EU.ImageManager.getTextureForKey(texturePath);
            if (!texture) {
                var message = "cannot create animation. Path [" + texturePath + "].";
                alert(message);
                continue;
            }
            var rect = cc.rect(0, 0, 0, 0);
            rect.width = texture.getContentSize().width;
            rect.height = texture.getContentSize().height;

            frames.push(new cc.SpriteFrame(texture, rect));
        }
        var delay = duration / frames.length;

        return new cc.Animation(frames, delay);
    }
};